File: python_version/parsing.py
from re import compile
from pathlib import Path
from typing import TextIO

from exceptions import (
    EmptyFileError,
    InvalidActionError,
    InvalidPlateauSizeError,
    InvalidRoverParamsError,
    MissingActionsError,
)
from models import Direction, Action, Plateau, Rover

# matches two non-negative integers separated by a space e.g. "5 5"
PLATEAU_RE = compile(r"^\d+ \d+$")
# matches two integers and a cardinal direction separated by spaces e.g. "1 2 N"
ROVER_RE = compile(rf"^-?\d+ -?\d+ ({'|'.join(d.value for d in Direction)})$")


def parse_plateau(line: str) -> Plateau:
    if not line:
        raise EmptyFileError

    if not PLATEAU_RE.fullmatch(line):
        raise InvalidPlateauSizeError

    x, y = line.strip().split(" ")
    return Plateau(max_x=int(x), max_y=int(y))


def parse_rover(line: str) -> Rover:
    if not ROVER_RE.fullmatch(line):
        raise InvalidRoverParamsError

    x, y, direction = line.strip().split(" ")
    return Rover(
        x_coordinate=int(x),
        y_coordinate=int(y),
        direction=Direction(direction),
    )


def parse_actions(line: str) -> list[Action]:
    actions = []
    for character in line.strip():
        try:
            actions.append(Action(character))
        except ValueError as e:
            raise InvalidActionError from e
    return actions


def extract_rover(*, plateau: Plateau, line: str) -> Rover:
    rover = parse_rover(line)

    if not plateau.contains(rover.x_coordinate, rover.y_coordinate):
        raise InvalidRoverParamsError

    return rover


def iter_non_empty_lines(f: TextIO):
    for line in f:
        if stripped := line.strip():
            yield stripped


def parse_stream(f: TextIO) -> tuple[Plateau, list[tuple[Rover, list[Action]]]]:
    lines = iter_non_empty_lines(f)
    plateau = parse_plateau(next(lines, ""))

    rovers_w_actions = []

    for line in lines:
        rover = extract_rover(plateau=plateau, line=line)

        actions_definition = next(lines, "")
        if not actions_definition:
            raise MissingActionsError(str(rover))

        actions = parse_actions(actions_definition)
        rovers_w_actions.append((rover, actions))

    return plateau, rovers_w_actions


def parse_file(path: str | Path) -> tuple[Plateau, list[tuple[Rover, list[Action]]]]:
    with open(Path(path)) as f:
        return parse_stream(f)

File: go_version/parsing.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var (
	// matches two non-negative integers separated by a space e.g. "5 5"
	plateauRE = regexp.MustCompile(`^\d+ \d+$`)
	// matches two integers and a cardinal direction separated by spaces e.g. "1 2 N"
	roverRE = regexp.MustCompile(`^-?\d+ -?\d+ [NESW]$`)
)

func parsePlateau(line string) (*Plateau, error) {
	if line == "" {
		return nil, ErrEmptyFile
	}
	if !plateauRE.MatchString(line) {
		return nil, ErrInvalidPlateauSize
	}
	xString, yString, _ := strings.Cut(line, " ") // regex guarantees valid non-negative integers
	x, _ := strconv.Atoi(xString)
	y, _ := strconv.Atoi(yString)
	return NewPlateau(x, y)
}

func parseRover(line string) (Rover, error) {
	if !roverRE.MatchString(line) {
		return Rover{}, ErrInvalidRoverParams
	}
	xString, rest, _ := strings.Cut(line, " ") // regex guarantees valid integers
	yString, direction, _ := strings.Cut(rest, " ")
	x, _ := strconv.Atoi(xString)
	y, _ := strconv.Atoi(yString)
	return Rover{X: x, Y: y, Direction: Direction(direction)}, nil
}

func parseActions(line string) ([]Action, error) {
	actions := make([]Action, 0, len(line))
	for _, ch := range line {
		a := Action(string(ch))
		switch a {
		case Left, Right, Forward:
			actions = append(actions, a)
		default:
			return nil, fmt.Errorf("%w: %c", ErrInvalidAction, ch)
		}
	}
	return actions, nil
}

func extractRover(plateau *Plateau, line string) (Rover, error) {
	rover, err := parseRover(line)
	if err != nil {
		return Rover{}, err
	}
	if !plateau.Contains(rover.X, rover.Y) {
		return Rover{}, ErrInvalidRoverParams
	}
	return rover, nil
}

type RoverWithActions struct {
	Rover   Rover
	Actions []Action
}

func parseStream(r io.Reader) (*Plateau, []RoverWithActions, error) {
	scanner := bufio.NewScanner(r)
	var lines []string
	for scanner.Scan() {
		if line := strings.TrimSpace(scanner.Text()); line != "" {
			lines = append(lines, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, nil, err
	}

	if len(lines) == 0 {
		return nil, nil, ErrEmptyFile
	}

	plateau, err := parsePlateau(lines[0])
	if err != nil {
		return nil, nil, err
	}

	var result []RoverWithActions
	for i := 1; i < len(lines); i += 2 {
		rover, err := extractRover(plateau, lines[i])
		if err != nil {
			return nil, nil, err
		}

		if i+1 >= len(lines) {
			return nil, nil, MissingActionsError{Rover: rover.Coordinates()}
		}

		actions, err := parseActions(lines[i+1])
		if err != nil {
			return nil, nil, err
		}

		result = append(result, RoverWithActions{Rover: rover, Actions: actions})
	}

	return plateau, result, nil
}

func parseFile(path string) (*Plateau, []RoverWithActions, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	defer func() { _ = f.Close() }()
	return parseStream(f)
}
